### ISSUE TITLE
App crashes when setLauncherVisibility called using v3.1.1

### ISSUE THREAD

**Issue by apo1981** [intercom/intercom-android#325]
## Version info
  - Intercom for Android version: 3.1.1
  - Android OS version: 6.0.1
  - Android Device: Samsung Galaxy S6

## Expected behavior
Application must correctly open showing the badge when unread conversations
## Actual behavior
Application crashes when setLauncherVisibility method called conversations

## Logs

03-23 15:33:48.398 3350-3350/? W/Intercom realtime: No realtime endpoints present so we can't connect
03-23 15:33:48.448 3350-3495/? W/System.err: remove failed: ENOENT (No such file or directory) : /data/user/0/com.example.app/cache/Intercom_SDK/HttpCache/journal.tmp
03-23 15:33:49.198 3350-3350/? E/AndroidRuntime: FATAL EXCEPTION: main
                                                 Process: com.example.app, PID: 3350
                                                 java.lang.StackOverflowError: stack size 8MB
                                                     at io.intercom.android.sdk.store.Store$3.dispatch(Store.java:74)
                                                     at io.intercom.android.sdk.middleware.LoggerMiddleware.dispatch(LoggerMiddleware.java:17)
                                                     at io.intercom.android.sdk.store.Store$3.dispatch(Store.java:74)
                                                     at io.intercom.android.sdk.store.Store.dispatch(Store.java:81)
                                                     at io.intercom.android.sdk.RealIntercom.setLauncherVisibility(RealIntercom.java:322)
                                                     at com.example.app.container.ContainerActivity$12.onCountUpdate(ContainerActivity.java:1071)
                                                     at io.intercom.android.sdk.store.UnreadCountTracker$1.onStateChange(UnreadCountTracker.java:28)
                                                     at io.intercom.android.sdk.store.UnreadCountTracker$1.onStateChange(UnreadCountTracker.java:26)
                                                     at io.intercom.android.sdk.store.ChangeSubscriber$1.run(ChangeSubscriber.java:35)
                                                     at io.intercom.android.sdk.store.ChangeSubscriber.runOnMainThread(ChangeSubscriber.java:44)
                                                     at io.intercom.android.sdk.store.ChangeSubscriber.onStateChange(ChangeSubscriber.java:33)
                                                     at io.intercom.android.sdk.store.Store$1.dispatch(Store.java:52)
                                                     at io.intercom.android.sdk.store.Store$2.dispatch(Store.java:66)
                                                     at io.intercom.android.sdk.middleware.AudioMiddleware.dispatch(AudioMiddleware.java:31)
                                                     at io.intercom.android.sdk.store.Store$3.dispatch(Store.java:74)
                                                     at io.intercom.android.sdk.middleware.OverlayManagerMiddleware.dispatch(OverlayManagerMiddleware.java:71)
                                                     at io.intercom.android.sdk.store.Store$3.dispatch(Store.java:74)
                                                     at io.intercom.android.sdk.middleware.NexusClientMiddleware.dispatch(NexusClientMiddleware.java:23)
                                                     at io.intercom.android.sdk.store.Store$3.dispatch(Store.java:74)
                                                     at io.intercom.android.sdk.middleware.UserUpdateBatcherMiddleware.dispatch(UserUpdateBatcherMiddleware.java:24)
                                                     at io.intercom.android.sdk.store.Store$3.dispatch(Store.java:74)
                                                     at io.intercom.android.sdk.middleware.ApiMiddleware.dispatch(ApiMiddleware.java:20)
                                                     at io.intercom.android.sdk.store.Store$3.dispatch(Store.java:74)
                                                     at io.intercom.android.sdk.middleware.LoggerMiddleware.dispatch(LoggerMiddleware.java:17)
                                                     at io.intercom.android.sdk.store.Store$3.dispatch(Store.java:74)
                                                     at io.intercom.android.sdk.store.Store.dispatch(Store.java:81)
                                                     at io.intercom.android.sdk.RealIntercom.setLauncherVisibility(RealIntercom.java:322)
                                                     at com.example.app.container.ContainerActivity$12.onCountUpdate(ContainerActivity.java:1071)
                                                     at io.intercom.android.sdk.store.UnreadCountTracker$1.onStateChange(UnreadCountTracker.java:28)
                                                     at io.intercom.android.sdk.store.UnreadCountTracker$1.onStateChange(UnreadCountTracker.java:26)
                                                     at io.intercom.android.sdk.store.ChangeSubscriber$1.run(ChangeSubscriber.java:35)
                                                     at io.intercom.android.sdk.store.ChangeSubscriber.runOnMainThread(ChangeSubscriber.java:44)
                                                     at io.intercom.android.sdk.store.ChangeSubscriber.onStateChange(ChangeSubscriber.java:33)
                                                     at io.intercom.android.sdk.store.Store$1.dispatch(Store.java:52)
                                                     at io.intercom.android.sdk.store.Store$2.dispatch(Store.java:66)
                                                     at io.intercom.android.sdk.middleware.AudioMiddleware.dispatch(AudioMiddleware.java:31)
                                                     at io.intercom.android.sdk.store.Store$3.dispatch(Store.java:74)
                                                     at io.intercom.android.sdk.middleware.OverlayManagerMiddleware.dispatch(OverlayManagerMiddleware.java:71)
                                                     at io.intercom.android.sdk.store.Store$3.dispatch(Store.java:74)
                                                     at io.intercom.android.sdk.middleware.NexusClientMiddleware.dispatch(NexusClientMiddleware.java:23)
                                                     at io.intercom.android.sdk.store.Store$3.dispatch(Store.java:74)
                                                     at io.intercom.android.sdk.middleware.UserUpdateBatcherMiddleware.dispatch(UserUpdateBatcherMiddleware.java:24)
                                                     at io.intercom.android.sdk.store.Store$3.dispatch(Store.java:74)
                                                     at io.intercom.android.sdk.middleware.ApiMiddleware.dispatch(ApiMiddleware.java:20)
                                                     at io.intercom.android.sdk.store.Store$3.dispatch(Store.java:74)
                                                     at io.intercom.android.sdk.middleware.LoggerMiddleware.dispatch(LoggerMiddleware.java:17)
                                                     at io.intercom.android.sdk.store.Store$3.dispatch(Store.java:74)
                                                     at io.intercom.android.sdk.store.Store.dispatch(Store.java:81)
                                                 	at io.intercom.android.sdk.Rea

**Comment by bboyle18** [comment]
Hi @apo1981 👋 

Thanks for reporting this.
We'll look at getting this fixed for the next release.

Brian

**Comment by apo1981** [comment]
Let me share more information

compileSdkVersion 25
buildToolsVersion '25.0.2'

Gradle version: 2.3.0
Android Studio IDE: v2.3

and this is the piece of code used

```
public void onResume() {
      ...
      addIntercomUnreadConversationListener();
      ...
}

// Add listener for showing badge indicating number of Unread conversations
private void addIntercomUnreadConversationListener() {
       Intercom.client().addUnreadConversationCountListener(unreadConversationCountListener);
}

private final UnreadConversationCountListener unreadConversationCountListener = new UnreadConversationCountListener() {
        @Override public void onCountUpdate(int unreadCount) {
            Intercom.Visibility mLauncherVisibility;
            if (unreadCount > 0) {
                mLauncherVisibility = Intercom.Visibility.VISIBLE;
            } else {
                mLauncherVisibility = Intercom.Visibility.GONE;
            }
            Intercom.client().setLauncherVisibility(mLauncherVisibility);
        }
    };
```

**Comment by bboyle18** [comment]
Thanks @apo1981 

We've managed to identify the problem and are working on a fix.

Brian

**Comment by ronocod** [comment]
We'll have a fix for this in the next release. In the mean time you can work around this by replacing 
```java
Intercom.client().setLauncherVisibility(mLauncherVisibility);
```

with
```java
new Handler().post(new Runnable() {
    @Override public void run() {
        Intercom.client().setLauncherVisibility(mLauncherVisibility);
    }
});
```

**Comment by apo1981** [comment]
Thks so much. The workaround works as expected.

**Comment by ronocod** [comment]
This is fixed in 3.1.2, if you use that version you won't need the workaround.
